### ISSUE TITLE
Aggregating a dimension using the Quantiles method with `skipna=True` is very slow

### ISSUE THREAD

**Issue by maawoo** [pydata/xarray#7377]
### What happened?

Hi all,
as the title already summarizes, I'm running into performance issues when aggregating over the time-dimension of a 3D DataArray using the [quantiles](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.quantile.html?highlight=quantiles) method with `skipna=True`. 
See the section below for some dummy data that represents what I'm working with (e.g., similar to [this](https://planetarycomputer.microsoft.com/dataset/sentinel-1-rtc)). Aggregating over the time-dimension of this dummy data I'm getting the following wall times: 

|   |   |  |
| --------------- | --------------- | --------------- | 
| 1 | `da.median(dim='time', skipna=True)` | 1.35 s |
| 2 | `da.quantile(0.95, dim='time', skipna=False)` | 5.95 s |
| 3 | `da.quantile(0.95, dim='time', skipna=True)` | 6 min 6s |

I'm currently using a compute node with 40 CPUs and 180 GB RAM. Here is what the resource utilization looks like. First small bump are 1 and 2. Second longer peak is 3.

![Screenshot 2022-12-14 at 17 33 14](https://user-images.githubusercontent.com/56583917/207654729-7ccecfc9-93f9-49f3-9bff-18f8643996d3.png)

In this small example, the process at least finishes after a few seconds. With my actual dataset the quantile calculation takes hours... 

I guess the following issue is relevant and should be revived:  https://github.com/numpy/numpy/issues/16575

Are there any possible work-arounds?

### Minimal Complete Verifiable Example

```Python
import pandas as pd
import numpy as np
import xarray as xr

# Create dummy data with 20% random NaNs
size_spatial = 2000
size_temporal = 20
n_nan = int(size_spatial**2*0.2)

time = pd.date_range("2000-01-01", periods=size_temporal)
lat = np.random.uniform(low=-90, high=90, size=size_spatial)
lon = np.random.uniform(low=-180, high=180, size=size_spatial)
data = np.random.rand(size_temporal, size_spatial, size_spatial)
index_nan = np.random.choice(data.size, n_nan, replace=False)
data.ravel()[index_nan] = np.nan

# Create DataArray
da = xr.DataArray(data=data, 
                  dims=['time', 'x', 'y'], 
                  coords={'time': time, 'x': lon, 'y': lat}, 
                  attrs={'nodata': np.nan})

# Calculate 95th quantile over time-dimension
da.quantile(0.95, dim='time', skipna=True)
```

### MVCE confirmation

- [x] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [x] Complete example — the example is self-contained, including all data and the text of any traceback.
- [ ] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [x] New issue — a search of GitHub Issues suggests this is not a duplicate.

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:36:39) [GCC 10.4.0]
python-bits: 64
OS: Linux
OS-release: 5.4.0-125-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: 4.9.0

xarray: 2022.12.0
pandas: 1.5.0
numpy: 1.23.3
scipy: 1.9.1
netCDF4: 1.6.1
pydap: None
h5netcdf: None
h5py: 3.7.0
Nio: None
zarr: None
cftime: 1.6.2
nc_time_axis: None
PseudoNetCDF: None
rasterio: 1.3.3
cfgrib: None
iris: None
bottleneck: 1.3.5
dask: 2022.10.0
distributed: 2022.10.0
matplotlib: 3.6.1
cartopy: 0.21.0
seaborn: 0.12.0
numbagg: None
fsspec: 2022.8.2
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 65.5.0
pip: 22.3
conda: 4.12.0
pytest: None
mypy: None
IPython: 8.5.0
sphinx: None
</details>

**Comment by maawoo** [comment]
This issue has an extra layer of evilness because users will also run into this issue when they don't specify the `skipna` parameter and their data is a float dtype, like in my example dummy data: `da.quantile(0.95, dim='time')`

The documentation could be a little bit clearer. The fact that `skipna=True` is the default for float dtypes could easily be overlooked in my opinion:

>  If True, skip missing values (as marked by NaN). By default, only skips missing values for float dtypes;

**Comment by arongergely** [comment]
Hi, this is a known issue coming from numpy.nanquantile / numpy.nanpercentile.
I had the same problem - AFAIK the workaround is to implement your own nanpercentiles calculation.

If you want to go that route:

There is a blog post about the issue + a numpy workaround for 3D arrays:
https://krstn.eu/np.nanpercentile()-there-has-to-be-a-faster-way/

I also turned to the numpy mailing list. Abel Aoun had a suggestion to look into the algo used at the [xclim](https://xclim.readthedocs.io/en/stable/index.html) project.  See our thread here here: https://mail.python.org/archives/list/numpy-discussion@python.org/message/EKQIS4KNOHS6ZAU5OSYTLNOOH7U2Y5TW/

I ended up taking that one and rewrote it to suit my needs. I achieved >100x speedup in my case
Good luck!

**Comment by maawoo** [comment]
Thanks @arongergely!
I have mentioned the numpy issue in my post above (FYI, for anyone looking for it). I was really surprised to see that it's over 2 years old and that this is now the first Xarray issue referencing it. If it's really a "well known" issue, I think it should have been somehow mentioned in the Xarray quantiles method.

I have seen the blog post and tried to use the workaround with apply_ufunc and Dask but ran into some problems. I'll revisit that when I have some time and will also check xclim. Seems to be very promising, thanks!

Happy holidays! 🎄

**Comment by dcherian** [comment]
Updates to the docstring adding all this information would be very welcome!

**Comment by maawoo** [comment]
Hi all,
I just created a simple workaround, which might be useful for others:  https://gist.github.com/maawoo/0b34d371c3cc1960a1589ccaded868c2

It uses the `_nan_quantile` method of xclim and works fine for my applications. Here is a quick comparison using the same example data as in my initial post:
<img width="592" alt="Screenshot 2023-01-27 at 13 41 42" src="https://user-images.githubusercontent.com/56583917/215089741-0504c4d7-9b62-4540-b61d-a59bfedf9741.png">

**Comment by zoj613** [comment]
> 

Does this work for an array of quantiles and also does it require the time coordinate to have a single chunk?

**Comment by lbferreira** [comment]
Recently, I create a library for the same purpose, maybe it can help other people. It can speed up quantile computations a lot.
To install, just run: `pip install fastnanquantile`
It's compatible with dask.
More info: https://github.com/lbferreira/fastnanquantile

Example below:
```python

import numpy as np
import xarray as xr
from fastnanquantile import xrcompat

da = xr.DataArray(
    np.random.rand(10, 1000, 1000),
    coords={"time": np.arange(10), "x": np.arange(1000), "y": np.arange(1000)},
)

# Xarray quantile (time to run: ~25s)
result_xr = da.quantile(q=0.6, dim="time")
# fastnanquantile (time to run: <1s)
result_fnq = xrcompat.xr_apply_nanquantile(da, q=0.6, dim="time")
# Check if results are equal (If results are different, an error will be raised)
np.testing.assert_almost_equal(result_fnq.values, result_xr.values, decimal=4)
```

**Comment by dcherian** [comment]
Thanks @lbferreira Do you know if its any faster than numbagg's nanquantile: https://github.com/numbagg/numbagg/issues/161 .

**Comment by lbferreira** [comment]
@dcherian My implementation is also based on numba. I just compared fastnanquantile against numbagg.
In a single benchmark (below) numbagg was faster (~5x). However, numbagg used 100% of my CPU (AMD Ryzen 9 7950X 16-core (32 threads), while fastnanquantile was not executed in parallel (CPU use around 5%). Thus, using fastnanquantile with Xarray+Dask may be a faster solution.

```python
import numpy as np
import numbagg
import fastnanquantile as fnq

sample_data = np.random.random((50, 100, 100))
%timeit np_result = np.nanquantile(sample_data, q=0.6, axis=0)
%timeit fnq_result = fnq.nanquantile(sample_data, q=0.6, axis=0)
%timeit nagg_result = numbagg.nanquantile(sample_data, quantiles=0.6, axis=0)

# Output 
# 269 ms ± 1.57 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# 9 ms ± 17.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# 1.68 ms ± 17.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Comment by max-sixty** [comment]
Yeah, having it run in parallel is going to generally be much faster. And avoiding multiple processes / dask is going to generally save on overhead.

A welcome PR would be to add numbagg's `nanpercentile` to xarray, in the same way that rolling methods run through numbagg by default if it's installed.

(or if we had benchmarks showing another approach was more performant, happy to consider that)

**Comment by max-sixty** [comment]
That wrapper looks good @maawoo !

Integrating that into the existing Xarray quantiles code would be a nice contribution if you're up for it...

**Comment by maawoo** [comment]
I'll give it a go @max-sixty 👍🏻

**Comment by lbferreira** [comment]
Just a point to think… calling a method (xarray.quantile) that always use all cpu cores (method backed with numbagg) may not be an expected behavior from an user perspective. I don’t know if numbagg provides some way to control whether to use multiple cpu cores. I just noted that it used all my cpu cores when I tested it.

**Comment by maawoo** [comment]
AFAIK, performance increasing backends such as `numbagg` and `bottleneck` are optional dependencies. So by default `numpy.nanquantile` will be used. In cases where users have `numbagg` installed but want to disable it, they could do so using [`xarray.set_options`](https://docs.xarray.dev/en/stable/generated/xarray.set_options.html):

```python
with xr.set_options(use_numbagg=False):
    # do something
```

**Comment by max-sixty** [comment]
> I'll give it a go @max-sixty 👍🏻

Great! Some of the existing code is great quality, some less so. So feel free to open up a PR with a draft of whatever you get done in an hour even if it's only partially complete, happy to help with questions

**Comment by dcherian** [comment]
Thanks @lbferreira I'd follow the approach we take here:
https://github.com/pydata/xarray/blob/614c25b165c18458570f5a68e03034f9b87cf1e7/xarray/core/nputils.py#L287

and then import that new `nanquantile` in the existing places we use the numpy method
